src/rebuild/srv.c

```diff
@@ -957,20 +962,20 @@ rebuild_scan_broadcast(struct ds_pool *pool, struct rebuild_global_pool_tracker
 		rc = rso->rso_status;
 
 	rgt->rgt_init_scan = 1;
-	if (rgt->rgt_stable_epoch == 0) {
-		rgt->rgt_stable_epoch = rso->rso_stable_epoch;
-	} else {
-		/* If initial rebuild/reintegration failed, the retry will keep
-		 * the same stable epoch to make sure reclaim only delete the data
-		 * before stable epoch, but keep the in-flight data.
+	rgt->rgt_stable_epoch = rso->rso_stable_epoch;
+	if (rgt->rgt_reclaim_epoch == 0) {
+		/* Keep the reclaim epoch, so once rebuild fails, the in-flight
+		 * data will not be discarded during reclaim by this epoch.
+		 * NB: the reclaim epoch should be initial rebuild job stable
+		 * epoch, to avoid discarding the in-flight data since the initial
+		 * rebuild.
 		 */
-		D_ASSERTF(rgt->rgt_stable_epoch <= rso->rso_stable_epoch,
-			  "stable epoch "DF_X64" > collect epoch "DF_X64"\n",
-			  rgt->rgt_stable_epoch, rso->rso_stable_epoch);
+		rgt->rgt_reclaim_epoch = rgt->rgt_stable_epoch;
```

Comment:
This is confusing.  In the PR comment, you say the stable epoch remains the same implying that the reclaim epoch changes.  But it looks like here, the reclaim epoch is the original stable epoch?

Reply:
reclaim epoch will only be assigned as stable epoch in the initial rebuild job, and if the initial rebuild job failed, then it needs to retry the rebuild, which it will keep the original reclaim epoch, instead of using the new stable epoch.  So it will only assign stable epoch if reclaim epoch is 0.

Reply:
so after this change, the retried rebuild will always use updated higher rgt_stable_epoch, but "rgt->rgt_reclaim_epoch" still keep the first-time rebuild's lower epoch.

Sorry I still feel some confuse about the original problem this patch try to resolve, the reclaim only discard the newly rebuilt object on the target of failed rebuild, will not discard the normal data (which rebuild fetch from), so why the "migration fetch with old stable epoch might missing some data, then causing data corruption"? because even if the migration fetch with old stable epoch should not read from those discarded data right?

Reply:
> so after this change, the retried rebuild will always use updated rgt_stable_epoch (higher), but "rgt->rgt_reclaim_epoch" still keep the first-time rebuild's stable epoch (lower).

yes, exactly.

> 
> Sorry I still feel some confuse about the original problem this patch try to resolve, the reclaim only discard the newly rebuilt object on the target of failed rebuild, will not discard the normal data (which rebuild fetch from), so why the "migration fetch with old stable epoch might missing some data, then causing data corruption"? because even if the migration fetch with old stable epoch should not read from those discarded data right?

If rebuild fail, it should only discard the data rebuilt from the rebuild, but keep the in-flight update data during rebuild, i.e. the data after reclaim epoch( == stable epoch of the initial rebuild).   Then during retry it should keep the original reclaim epoch,   otherwise it may discard the in-flight data from the initial rebuild.    

The issue is that during retry, we should use the new stable epoch, instead of the old stable epoch from the original rebuild, which will be used to fetch the rebuild data from other targets,  if we are using the older stable epoch, some data will be missed, especially if the retry might merge the new failure targets.  So the patch separate the reclaim epoch and stable epoch here.

> 
> Just thought another possible problem, for the first rebuild "rgt->rgt_stable_epoch = rso->rso_stable_epoch" that is highest hlc on all nodes, that is higher than local HLC on some nodes. So the inflight update possibly will on those nodes with epoch < rgt_stable_epoch. Then if the rebuild failed on those node, it reclaim/discard the data by using rgt_stable_epoch, that may discard newly updated data. Is it a problem?

hmm, I think you are right. I will resolve it in a different patch. Thanks!